[dynamo] add side_effects logging artifact

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #171469

Implements https://github.com/pytorch/pytorch/issues/169722

Added `TORCH_LOGS="side_effects"` to track which side effects Dynamo codegen'd.

This information to tlparse as well. Sample output in tlparse:
```
Mutating object of type list (source name: G['lst'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 8, in fn
        lst.append(4)

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 9, in fn
        lst.append(5)


========================================

Mutating object of type dict (source name: G['dct'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 10, in fn
        dct[3] = 4

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 11, in fn
        dct[4] = 5

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 12, in fn
        dct[5] = 6
```

Also modifies `torch._dynamo.config.side_effect_replay_policy`:
- `"info"` (default) logs all side effects codegen'd by Dynamo
- `"silent"` does not log side effects but still codegens them. Use if `"info"` is causing compile time problems due to excessive logging.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @mlazos

williamwen42 committed:
[dynamo] add side_effects logging artifact

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add side_effects logging artifact"


Implements https://github.com/pytorch/pytorch/issues/169722

Added `TORCH_LOGS="side_effects"` to track which side effects Dynamo codegen'd.

Follow-up: we could log this information to tlparse as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add side_effects logging artifact"


Implements https://github.com/pytorch/pytorch/issues/169722

Added `TORCH_LOGS="side_effects"` to track which side effects Dynamo codegen'd.

Follow-up: we could log this information to tlparse as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

Lucaskabela reviewed (commented):
Minor comment on the logging, but the change generally looks good 

Will wait for thoughts on this change as well as a rebase with working CI (right now it is failing so can't get useful signals)

williamwen42 commented:
Since side effects logging happens unconditionally, we don't need to add tests for logging side effects on each VariableTracker subtype - existing side effects tests on those should be sufficient coverage.

williamwen42 committed:
Update on "[dynamo] add side_effects logging artifact"


Implements https://github.com/pytorch/pytorch/issues/169722

Added `TORCH_LOGS="side_effects"` to track which side effects Dynamo codegen'd.

Follow-up: we could log this information to tlparse as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add side_effects logging artifact"


Implements https://github.com/pytorch/pytorch/issues/169722

Added `TORCH_LOGS="side_effects"` to track which side effects Dynamo codegen'd.

This information to tlparse as well. Sample output in tlparse:
```
Mutating object of type list (source name: G['lst'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 8, in fn
        lst.append(4)

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 9, in fn
        lst.append(5)


========================================

Mutating object of type dict (source name: G['dct'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 10, in fn
        dct[3] = 4

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 11, in fn
        dct[4] = 5

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 12, in fn
        dct[5] = 6
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add side_effects logging artifact"


Implements https://github.com/pytorch/pytorch/issues/169722

Added `TORCH_LOGS="side_effects"` to track which side effects Dynamo codegen'd.

This information to tlparse as well. Sample output in tlparse:
```
Mutating object of type list (source name: G['lst'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 8, in fn
        lst.append(4)

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 9, in fn
        lst.append(5)


========================================

Mutating object of type dict (source name: G['dct'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 10, in fn
        dct[3] = 4

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 11, in fn
        dct[4] = 5

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 12, in fn
        dct[5] = 6
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

Lucaskabela commented:
Any thoughts on why these benchmarks are regressing? Would we expect the logging to add a significant instruction overhead here?

Lucaskabela reviewed (changes requested):
SG - just rerequest review or ping when we have it fixed

williamwen42 committed:
Update on "[dynamo] add side_effects logging artifact"


Implements https://github.com/pytorch/pytorch/issues/169722

Added `TORCH_LOGS="side_effects"` to track which side effects Dynamo codegen'd.

This information to tlparse as well. Sample output in tlparse:
```
Mutating object of type list (source name: G['lst'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 8, in fn
        lst.append(4)

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 9, in fn
        lst.append(5)


========================================

Mutating object of type dict (source name: G['dct'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 10, in fn
        dct[3] = 4

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 11, in fn
        dct[4] = 5

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 12, in fn
        dct[5] = 6
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 commented:
Deferred full stack extraction + conversion to string until logging, which seems to make the compile time microbenchmark regressions go away.

williamwen42 committed:
Update on "[dynamo] add side_effects logging artifact"


Implements https://github.com/pytorch/pytorch/issues/169722

Added `TORCH_LOGS="side_effects"` to track which side effects Dynamo codegen'd.

This information to tlparse as well. Sample output in tlparse:
```
Mutating object of type list (source name: G['lst'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 8, in fn
        lst.append(4)

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 9, in fn
        lst.append(5)


========================================

Mutating object of type dict (source name: G['dct'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 10, in fn
        dct[3] = 4

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 11, in fn
        dct[4] = 5

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 12, in fn
        dct[5] = 6
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

Lucaskabela reviewed (approved):
Thanks for change, LGTM

williamwen42 committed:
Update on "[dynamo] add side_effects logging artifact"


Implements https://github.com/pytorch/pytorch/issues/169722

Added `TORCH_LOGS="side_effects"` to track which side effects Dynamo codegen'd.

This information to tlparse as well. Sample output in tlparse:
```
Mutating object of type list (source name: G['lst'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 8, in fn
        lst.append(4)

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 9, in fn
        lst.append(5)


========================================

Mutating object of type dict (source name: G['dct'])

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 10, in fn
        dct[3] = 4

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 11, in fn
        dct[4] = 5

    ********

      File "/data/users/williamwen/pytorch4/playground.py", line 15, in <module>
        fn(torch.ones(3))
      File "/data/users/williamwen/pytorch4/playground.py", line 12, in fn
        dct[5] = 6
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
